### ISSUE TITLE
unique doc pids

### ISSUE THREAD

**Issue by emanjavacas** [INL/BlackLab-server#9]
Hi there,

I've notice this note in the blacklabserver overview:

> BlackLab Server will use the Lucene document id instead of a true persistent identifier if your corpus has no persistent identifier configured, but this is not recommended: Lucene document ids can change if you re-index or compact the index, so bookmarked URLs may not always return to the same information.

I'd like to be sure that my index doc ids won't change after reindexing, however I am not quite sure how to enforce this in my DocIndexer class. Any hints?

best!

**Comment by jan-niestadt** [comment]
Yes, I agree that needs clarification. When creating an index using the IndexTool, it looks for a file indextemplate.json in the input directory (or the parent directory of the input directory). If you specify "pidField" in this file (see http://inl.github.io/BlackLab/indexing-with-blacklab.html#index-struct), the field you specify will be used as the persistent document identifier by BlackLab Server.

If you don't want to have to re-index, you can also edit the indexstructure.json file in the index directory directly.

I'll update the documentation. Thanks for pointing this out!

**Comment by emanjavacas** [comment]
Wow, I hadn't seen that documentation page before. Thanks for the link and for the info!
As far as I understand, the indexer will check through the xml document to find those fields like "author", "date", etc... I am still wondering what happens if there are more than one of those entries (e.g. nested at different levels of the document), or am I misunderstanding something?

**Comment by jan-niestadt** [comment]
Yes, I added the page about indexing fairly recently. I've been trying to improve the documentation lately, so by all means, keep asking questions if anything is unclear, then I can improve things.

To answer your question: it is up to the DocIndexer class to translate between the XML structure and the metadata fields. If the structure of the metadata is simple, you can use one of the predefined handlers like MetadataElementHandler, MetadataAttributesHandler or MetadataNameValueAttributeHandler; if the structure doesn't fit these default handlers, you should write your own. Have a look at [nl.inl.blacklab.index.DocIndexerXmlHandlers.java](https://github.com/INL/BlackLab/blob/master/core/src/main/java/nl/inl/blacklab/index/DocIndexerXmlHandlers.java) to see how the default ones are implemented, it's not difficult.

**Comment by jan-niestadt** [comment]
By the way, would you mind posting all questions to the [main BlackLab repository](https://github.com/INL/BlackLab/issues) from now on? I think it would be nice to archive them all in one place. This separate BlackLab Server repository will probably be removed at some point in the future now that the two projects are in one repository together. Thanks!

**Comment by emanjavacas** [comment]
yes, sorry, I just noticed right after posting that it was the wrong repo
